File: jobs/shared/src/db/events.ts
import pg from 'pg';
import { ResultAsync } from 'neverthrow';
import { databaseError, type DatabaseError } from '../errors.js';
import { getLogger } from '../logger.js';

export interface JobCompletedEvent {
  job_id: string;
  job_type: string;
  transactions_count: number;
  assets_count: number;
}

export type EventHandler<T> = (event: T) => Promise<void>;

export function listenForJobCompleted(
  pool: pg.Pool,
  handler: EventHandler<JobCompletedEvent>
): ResultAsync<() => Promise<void>, DatabaseError> {
  return ResultAsync.fromPromise(
    (async () => {
      const log = getLogger();
      const client = await pool.connect();

      client.on('notification', async (msg) => {
        if (msg.channel === 'mf_job_completed' && msg.payload) {
          try {
            const event = JSON.parse(msg.payload) as JobCompletedEvent;
            log.info({ event }, 'Received mf_job_completed notification');
            await handler(event);
          } catch (e) {
            log.error({ error: e, payload: msg.payload }, 'Failed to handle notification');
          }
        }
      });

      await client.query('LISTEN mf_job_completed');
      log.info('Listening for mf_job_completed events');

      return async () => {
        await client.query('UNLISTEN mf_job_completed');
        client.release();
        log.info('Stopped listening for mf_job_completed events');
      };
    })(),
    (e) => databaseError('Failed to listen for job completed events', e)
  );
}

export function notify(
  pool: pg.Pool,
  channel: string,
  payload: unknown
): ResultAsync<void, DatabaseError> {
  return ResultAsync.fromPromise(
    (async () => {
      await pool.query('SELECT pg_notify($1, $2)', [channel, JSON.stringify(payload)]);
    })(),
    (e) => databaseError(`Failed to notify channel ${channel}`, e)
  );
}

File: jobs/transform/src/transform/db/events.py
"""PostgreSQL LISTEN/NOTIFY event handling."""

import json
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

import psycopg
import structlog

log = structlog.get_logger()


@dataclass
class JobCompletedEvent:
    """Event emitted when a scraper job completes."""

    job_id: str
    job_type: str
    transactions_count: int
    assets_count: int


def listen_for_job_completed(
    conn_string: str,
    handler: Callable[[JobCompletedEvent], None],
) -> None:
    """Listen for mf_job_completed notifications and call handler.

    This is a blocking function that listens indefinitely.
    """
    with psycopg.connect(conn_string, autocommit=True) as conn:
        conn.execute("LISTEN mf_job_completed")
        log.info("Listening for mf_job_completed events")

        for notify in conn.notifies():
            try:
                payload: dict[str, Any] = json.loads(notify.payload) if notify.payload else {}
                event = JobCompletedEvent(
                    job_id=payload.get("job_id", ""),
                    job_type=payload.get("job_type", ""),
                    transactions_count=payload.get("transactions_count", 0),
                    assets_count=payload.get("assets_count", 0),
                )
                log.info("Received job completed event", event=event)
                handler(event)
            except json.JSONDecodeError as e:
                log.error("Failed to parse notification payload", error=str(e))
            except Exception as e:
                log.error("Error handling notification", error=str(e))
